### ISSUE TITLE
NCCL Timeout during the data processing step of training

### ISSUE THREAD

**Issue by RajenDey** [axolotl-ai-cloud/axolotl#921]
### Please check that this issue hasn't been reported before.

- [X] I searched previous [Bug Reports](https://github.com/OpenAccess-AI-Collective/axolotl/labels/bug) didn't find any similar reports.

### Expected Behavior

It should not timeout

### Current behaviour

Preprocessing the dataset does not save the sample packed dataset to disk. This causes the computation to be run again at the start of training, and if it exceeds 30 minutes (if the dataset is ~5gb or more) it causes an NCCL timeout error.

Changing ddp_timeout doesnt work, and neither does setting:
os.environ['NCCL_BLOCKING_WAIT'] = '1'
os.environ['NCCL_SOCKET_NTHREADS'] = '4'
os.environ['NCCL_NSOCKS_PERTHREAD'] = '8'
os.environ['NCCL_ASYNC_ERROR_HANDLING'] = '1'
os.environ['NCCL_P2P_LEVEL'] = 'NVL'

### Steps to reproduce

Run training or preprocessing then training with sample_packing=true on a large dataset.
Ex:
1. !python -m axolotl.cli.preprocess mine/llama-2-32k.yml --debug
2. !accelerate launch -m axolotl.cli.train mine/llama-2-32k.yml

### Config yaml

base_model: togethercomputer/LLaMA-2-7B-32K
model_type: LlamaForCausalLM
tokenizer_type: LlamaTokenizer
is_llama_derived_model: true

load_in_8bit: false
load_in_4bit: false
strict: false

datasets:
  - path: data/train/discord_test.jsonl
    type: completion
dataset_prepared_path: last_run_prepared
val_set_size: 0.05
output_dir: ./out3

sequence_len: 4096
sample_packing: true
pad_to_sequence_len: true

adapter:
lora_model_dir:
lora_r:
lora_alpha:
lora_dropout:
lora_target_linear:
lora_fan_in_fan_out:

wandb_project: league
wandb_entity: vijnej
wandb_watch:
wandb_run_id: run3
wandb_log_model:

gradient_accumulation_steps: 1
micro_batch_size: 1
num_epochs: 1
optimizer: lion_8bit
lr_scheduler: cosine
learning_rate: 0.00002

train_on_inputs: false
group_by_length: false
bf16: true
fp16: false
tf32: false

gradient_checkpointing: true
early_stopping_patience:
resume_from_checkpoint:
local_rank:
logging_steps: 1
xformers_attention:
flash_attention: true
flash_attn_cross_entropy: false
flash_attn_rms_norm: true
flash_attn_fuse_qkv: false
flash_attn_fuse_mlp: true

warmup_steps: 100
eval_steps: 0.05
eval_table_size:
save_steps: 0.1
debug:
deepspeed: deepspeed/zero2.json # multi-gpu only
weight_decay: 0.1
fsdp:
fsdp_config:
special_tokens:
  bos_token: "<s>"
  eos_token: "</s>"
  unk_token: "<unk>"
ddp_timeout: 9999

### Possible solution

save the processed dataset to disk after sample packing and load it in during train instead of recomputing it.

### Which Operating Systems are you using?

- [X] Linux
- [ ] macOS
- [ ] Windows

### Python Version

3.10

### axolotl branch-commit

main/3e3229e

### Acknowledgements

- [X] My issue title is concise, descriptive, and in title casing.
- [X] I have searched the existing issues to make sure this bug has not been reported yet.
- [X] I am using the latest version of axolotl.
- [X] I have provided enough information for the maintainers to reproduce and diagnose the issue.

**Comment by NanoCode012** [comment]
> Preprocessing the dataset does not save the sample packed dataset to disk. This causes the computation to be run again at the start of training, and if it exceeds 30 minutes (if the dataset is ~5gb or more) it causes an NCCL timeout error.

Hello, could you try set the following config?

```yml
dataset_prepared_path: last_run_prepared
```

The current example files do not load from cache. Setting the above will.

Secondly, could you also try to preprocess it first since the dataset is large? https://github.com/OpenAccess-AI-Collective/axolotl#preprocess-dataset

**Comment by ajinkya123-robo** [comment]
Increasing 

> val_set_size 

solved my problem. Plus 

> dataset_prepared_path: last_run_prepared

 can be handy.
